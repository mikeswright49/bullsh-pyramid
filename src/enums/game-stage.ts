export enum GameStage {
    Initiation,
    Memorization,
    Flipping,
    Declaration,
    Bullshit,
    Memory,
    Complete,
    Reveal,
}
