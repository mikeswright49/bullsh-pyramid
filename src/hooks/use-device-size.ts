import { useEffect, useState } from 'react';
import { debounce } from 'lodash';
/**
 * Enum for breakpoint values.
 * @readonly
 * @enum {number}
 */
export enum Breakpoint {
    XS = 0,
    SM = 575,
    MD = 768,
    LG = 992,
    XL = 1200,
}

type Handler = () => void;
let handlers: Handler[] = [];

const onResize = () => {
    handlers.forEach((handler) => {
        handler();
    });
};

/**
 *
 * export @function getCurrentBrakpoint(window.innerWidth)
 * @param {number} innerWidth
 *
 * * @example
 * // Get current breakpoint by passing @param {number} innerWidth
 * getCurrentBrakpoint(window.innerWidth);
 * @returns {enum} breakpoint value base on condition from value passed by
 * @param {number} innerWidth
 * getCurrentBrakpoint(window.innerWidth);
 * @param innerWidth — *
 * @param innerWidth — getCurrentBrakpoint(window.innerWidth);
 * @param innerWidth
 */

export function getCurrentBreakPoint(innerWidth: number): Breakpoint {
    if (innerWidth >= Breakpoint.XL) {
        return Breakpoint.XL;
    } else if (innerWidth >= Breakpoint.LG) {
        return Breakpoint.LG;
    } else if (innerWidth >= Breakpoint.MD) {
        return Breakpoint.MD;
    } else if (innerWidth >= Breakpoint.SM) {
        return Breakpoint.SM;
    }
    return Breakpoint.XS;
}

useDeviceSize.debounceDelay = 50;
useDeviceSize.initialized = false;
/**
 * React hook to get user device height, width, and breakpoint
 * @param {object} windowDI
 * @property {number}  innerHeight
 * @property {number}  innerWidth
 * @property {number}  breakpoint
 * @returns {object} innerHeight, innerWidth, breakpoint
 *
 * @example
 *
 * // Get the height, width, or breakpoint for the user device
 * const { height, width, breakpoint } = useDeviceSize();
 *
 * // Use returned object to display component when on designated height , width, or breakpoint
 * {height > 600 ? <Mobile/> : <Desktop/>}
 * {width > 768 ? <Tablet/> : <Desktop/>}
 * {breakpoint > 1200 ? <Desktop/> : <Mobile/>}
 *
 */

// tslint:disable-next-line: completed-docs
export function useDeviceSize(windowDI?: {
    innerHeight: number;
    innerWidth: number;
    addEventListener?: (key: string, callback: Handler) => void;
}): { height: number; width: number; breakpoint: Breakpoint } {
    const display = windowDI || window;
    const [deviceSize, setDeviceSize] = useState(() => ({
        height: display.innerHeight,
        width: display.innerWidth,
        breakpoint: getCurrentBreakPoint(display.innerWidth),
    }));

    useEffect(() => {
        // static singular registration of the useDeviceSize on all platforms
        if (!useDeviceSize.initialized) {
            (display as Window).addEventListener(
                'resize',
                debounce(onResize, useDeviceSize.debounceDelay)
            );
            useDeviceSize.initialized = true;
        }

        const register = (handler: Handler) => {
            handlers.push(handler);
        };

        const unregister = (handler: Handler) => {
            handlers = handlers.filter((existingHandler) => handler !== existingHandler);
        };

        const updateDeviceSize = () => {
            setDeviceSize({
                height: display.innerHeight,
                width: display.innerWidth,
                breakpoint: getCurrentBreakPoint(display.innerWidth),
            });
        };

        register(updateDeviceSize);
        return () => {
            unregister(updateDeviceSize);
        };
    }, []);
    return deviceSize;
}
