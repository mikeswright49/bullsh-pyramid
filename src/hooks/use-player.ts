import { Player } from 'types/player';
import { useState, useEffect } from 'react';
import { PlayerStore } from 'src/stores/player-store';

export function usePlayer(playerId: string) {
    const [player, setPlayer] = useState<Player>();

    useEffect(() => {
        if (!playerId) {
            return;
        }

        PlayerStore.subscribeToPlayer(playerId, (updatedPlayer) => {
            setPlayer(updatedPlayer);
        });

        return () => PlayerStore.unsubscribeToPlayer();
    }, [playerId]);

    return player;
}
