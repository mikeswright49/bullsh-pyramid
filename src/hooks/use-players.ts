import { useState, useEffect } from 'react';
import { GameStore } from 'src/stores/game-store';
import { Player } from 'types/player';
import { PlayerStore } from 'src/stores/player-store';

export function usePlayers(gameId: string) {
    const [players, setPlayers] = useState<{ [key: string]: Player }>({});

    useEffect(() => {
        if (!gameId) {
            return;
        }

        GameStore.subscribeToPlayers(gameId, (player) => {
            players[player.id] = player;
            PlayerStore.subscribeToHaters(player.id, (hater) => {
                const playerRef = players[player.id];
                if (!playerRef.haters) {
                    playerRef.haters = [];
                }
                playerRef.haters.push(hater);
                setPlayers({ ...players });
            });
            setPlayers({ ...players });
        });
    }, [gameId]);

    return players;
}
