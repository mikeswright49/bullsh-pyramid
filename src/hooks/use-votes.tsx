import { useState, useEffect } from 'react';
import { Vote } from 'types/vote';
import { VoteStore } from 'src/stores/vote-store';

export function useVotes(gameId: string) {
    const [votes, setVotes] = useState<{ [key: string]: Vote }>({});

    useEffect(() => {
        if (!gameId) {
            return;
        }

        VoteStore.subscribeToVotes(gameId, (vote: Vote) => {
            votes[vote.id] = vote;
            setVotes({ ...votes });
        });
    }, [gameId]);
    return votes;
}
