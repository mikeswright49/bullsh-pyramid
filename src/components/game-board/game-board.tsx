import React from 'react';
import { GameStage } from '../../enums/game-stage';
import { Pyramid } from '../pyramid/pyramid';
import { GameState } from 'types/game-state';
import styles from './game-board.module.css';

export function GameBoard(props: { gameState: GameState }) {
    const { gameState } = props;
    if (gameState.gameStage === GameStage.Initiation) {
        return <span data-testid="at-game-board"></span>;
    }
    return (
        <div className="flex-center-center" data-testid="at-game-board">
            <div className={styles.rightBar}>
                <Pyramid tiers={gameState.tiers} />
            </div>
        </div>
    );
}
