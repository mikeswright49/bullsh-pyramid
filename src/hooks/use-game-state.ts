import { useState, useEffect } from 'react';
import { GameState } from 'types/game-state';
import { GameStage } from '../enums/game-stage';
import { GameStore } from '../stores/game-store';

export const DEFAULT_GAME_STATE: GameState = {
    id: '',
    gameStage: GameStage.Initiation,
    activeRow: 0,
    activeIndex: 0,
    players: [],
    tiers: [],
};

export function useGameState(gameId: string) {
    const [gameState, setGameState] = useState<GameState>(DEFAULT_GAME_STATE);

    useEffect(() => {
        if (!gameId) {
            return;
        }

        GameStore.subscribeToGame(gameId, (updatedGameState) => {
            setGameState(updatedGameState);
        });

        return () => GameStore.unsubscribeToGame();
    }, [gameId]);

    return gameState;
}
