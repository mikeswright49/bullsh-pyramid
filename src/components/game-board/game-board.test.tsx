import React from 'react';
import { render } from '@testing-library/react';
import { GameBoard } from './game-board';
import { DEFAULT_GAME_STATE } from '../../hooks/use-game-state';
import { GameStage } from '../../enums/game-stage';

describe('<Unit Test> - Game board', () => {
    it('should return an empty span if in initialization state', () => {
        const { container } = render(<GameBoard gameState={DEFAULT_GAME_STATE} />);
        expect(container.querySelector('[data-testid="at-game-board"]')).toMatchSnapshot();
    });
    it('should render the pyramid if not in initialization', () => {
        const gameState = { ...DEFAULT_GAME_STATE };
        gameState.gameStage = GameStage.Bullshit;
        const { container } = render(<GameBoard gameState={gameState} />);

        expect(container.querySelector('[data-testid="at-game-board"]')).toMatchSnapshot();
    });
});
