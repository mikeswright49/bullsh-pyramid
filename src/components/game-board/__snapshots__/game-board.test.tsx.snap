// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Unit Test> - Game board should render the pyramid if not in initialization 1`] = `
<div
  class="flex-center-center"
  data-testid="at-game-board"
>
  <div
    class="rightBar"
  >
    <div
      data-testid="at-pyramid"
    />
  </div>
</div>
`;

exports[`<Unit Test> - Game board should return an empty span if in initialization state 1`] = `
<span
  data-testid="at-game-board"
/>
`;
